{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘tidyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:magrittr’:\n",
      "\n",
      "    extract\n",
      "\n",
      "\n",
      "Registered S3 methods overwritten by 'ggtern':\n",
      "  method           from   \n",
      "  grid.draw.ggplot ggplot2\n",
      "  plot.ggplot      ggplot2\n",
      "  print.ggplot     ggplot2\n",
      "\n",
      "--\n",
      "Remember to cite, run citation(package = 'ggtern') for further info.\n",
      "--\n",
      "\n",
      "\n",
      "Attaching package: ‘ggtern’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:ggplot2’:\n",
      "\n",
      "    aes, annotate, ggplot, ggplot_build, ggplot_gtable, ggplotGrob,\n",
      "    ggsave, layer_data, theme_bw, theme_classic, theme_dark,\n",
      "    theme_gray, theme_light, theme_linedraw, theme_minimal, theme_void\n",
      "\n",
      "\n",
      "Loading required package: viridisLite\n",
      "\n",
      "\n",
      "Attaching package: ‘arrow’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:magrittr’:\n",
      "\n",
      "    is_in\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    timestamp\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘purrr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:jsonlite’:\n",
      "\n",
      "    flatten\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    transpose\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:magrittr’:\n",
      "\n",
      "    set_names\n",
      "\n",
      "\n",
      "corrplot 0.92 loaded\n",
      "\n",
      "Loading required package: lattice\n",
      "\n",
      "\n",
      "Attaching package: ‘caret’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:purrr’:\n",
      "\n",
      "    lift\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    dcast, melt\n",
      "\n",
      "\n",
      "\n",
      "---------------------\n",
      "Welcome to dendextend version 1.17.1\n",
      "Type citation('dendextend') for how to cite the package.\n",
      "\n",
      "Type browseVignettes(package = 'dendextend') for the package vignette.\n",
      "The github page is: https://github.com/talgalili/dendextend/\n",
      "\n",
      "Suggestions and bug-reports can be submitted at: https://github.com/talgalili/dendextend/issues\n",
      "You may ask questions at stackoverflow, use the r and dendextend tags: \n",
      "\t https://stackoverflow.com/questions/tagged/dendextend\n",
      "\n",
      "\tTo suppress this message use:  suppressPackageStartupMessages(library(dendextend))\n",
      "---------------------\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘dendextend’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    set\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    cutree\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'GGally':\n",
      "  method from   \n",
      "  +.gg   ggplot2\n",
      "\n",
      "Registered S3 methods overwritten by 'fmsb':\n",
      "  method    from\n",
      "  print.roc pROC\n",
      "  plot.roc  pROC\n",
      "\n",
      "\n",
      "Attaching package: ‘patchwork’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:cowplot’:\n",
      "\n",
      "    align_plots\n",
      "\n",
      "\n",
      "Loading required package: Matrix\n",
      "\n",
      "\n",
      "Attaching package: ‘Matrix’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:tidyr’:\n",
      "\n",
      "    expand, pack, unpack\n",
      "\n",
      "\n",
      "Loading required package: arm\n",
      "\n",
      "Loading required package: MASS\n",
      "\n",
      "\n",
      "Attaching package: ‘MASS’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:patchwork’:\n",
      "\n",
      "    area\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "\n",
      "arm (Version 1.14-4, built: 2024-4-1)\n",
      "\n",
      "\n",
      "Working directory is /work_beegfs/sukmb667/projects/cdr3-qtl/healthy_and_ibd/scripts\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘arm’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:corrplot’:\n",
      "\n",
      "    corrplot\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘bestNormalize’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    boxcox\n",
      "\n",
      "\n",
      "Loading required package: parsnip\n",
      "\n",
      "\n",
      "Attaching package: ‘gghighlight’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:ggtern’:\n",
      "\n",
      "    aes\n",
      "\n",
      "\n",
      "Welcome! Want to learn more? See two factoextra-related books at https://goo.gl/ve3WBa\n",
      "\n",
      "\n",
      "Attaching package: ‘UpSetR’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lattice’:\n",
      "\n",
      "    histogram\n",
      "\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.18.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ComplexHeatmap’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:merTools’:\n",
      "\n",
      "    draw\n",
      "\n",
      "\n",
      "========================================\n",
      "circlize version 0.4.16\n",
      "CRAN page: https://cran.r-project.org/package=circlize\n",
      "Github page: https://github.com/jokergoo/circlize\n",
      "Documentation: https://jokergoo.github.io/circlize_book/book/\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. circlize implements and enhances circular visualization\n",
      "  in R. Bioinformatics 2014.\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(circlize))\n",
      "========================================\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'ggfortify':\n",
      "  method          from   \n",
      "  autoplot.glmnet parsnip\n",
      "\n",
      "Loading required package: Rcpp\n",
      "\n",
      "Loading 'brms' package (version 2.22.0). Useful instructions\n",
      "can be found by typing help('brms'). A more detailed introduction\n",
      "to the package is available through vignette('brms_overview').\n",
      "\n",
      "\n",
      "Attaching package: ‘brms’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:lme4’:\n",
      "\n",
      "    ngrps\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:stats’:\n",
      "\n",
      "    ar\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘mvabund’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:arm’:\n",
      "\n",
      "    coefplot\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘future’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:caret’:\n",
      "\n",
      "    cluster\n",
      "\n",
      "\n",
      "Classes and Methods for R originally developed in the\n",
      "Political Science Computational Laboratory\n",
      "Department of Political Science\n",
      "Stanford University (2002-2015),\n",
      "by and under the direction of Simon Jackman.\n",
      "hurdle and zeroinfl functions by Achim Zeileis.\n",
      "\n",
      "This is rstanarm version 2.32.1\n",
      "\n",
      "- See https://mc-stan.org/rstanarm/articles/priors for changes to default priors!\n",
      "\n",
      "- Default priors may change, so it's safest to specify priors, even if equivalent to the defaults.\n",
      "\n",
      "- For execution on a local, multicore CPU with excess RAM we recommend calling\n",
      "\n",
      "  options(mc.cores = parallel::detectCores())\n",
      "\n",
      "\n",
      "Attaching package: ‘rstanarm’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:brms’:\n",
      "\n",
      "    dirichlet, exponential, get_y, lasso, ngrps\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:arm’:\n",
      "\n",
      "    invlogit, logit\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:caret’:\n",
      "\n",
      "    compare_models, R2\n",
      "\n",
      "\n",
      "Loading required package: vegan\n",
      "\n",
      "Loading required package: permute\n",
      "\n",
      "\n",
      "Attaching package: ‘permute’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dendextend’:\n",
      "\n",
      "    shuffle\n",
      "\n",
      "\n",
      "Registered S3 methods overwritten by 'vegan':\n",
      "  method          from      \n",
      "  rev.hclust      dendextend\n",
      "  print.nullmodel parsnip   \n",
      "\n",
      "This is vegan 2.6-8\n",
      "\n",
      "\n",
      "Attaching package: ‘vegan’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:parsnip’:\n",
      "\n",
      "    nullmodel\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:caret’:\n",
      "\n",
      "    tolerance\n",
      "\n",
      "\n",
      "Registered S3 methods overwritten by 'vegan3d':\n",
      "  method            from \n",
      "  plot.orditkplot   vegan\n",
      "  points.orditkplot vegan\n",
      "  scores.orditkplot vegan\n",
      "  text.orditkplot   vegan\n",
      "\n",
      "\n",
      "Attaching package: ‘vegan3d’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:vegan’:\n",
      "\n",
      "    orditkplot\n",
      "\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ───────────────────────────────────────────── tidymodels 1.2.0 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mbroom       \u001b[39m 1.0.7     \u001b[32m✔\u001b[39m \u001b[34mtibble      \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdials       \u001b[39m 1.3.0     \u001b[32m✔\u001b[39m \u001b[34mtune        \u001b[39m 1.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34minfer       \u001b[39m 1.0.7     \u001b[32m✔\u001b[39m \u001b[34mworkflows   \u001b[39m 1.1.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mmodeldata   \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mworkflowsets\u001b[39m 1.1.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mrecipes     \u001b[39m 1.1.0     \u001b[32m✔\u001b[39m \u001b[34myardstick   \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mrsample     \u001b[39m 1.2.1     \n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ──────────────────────────────────────────────── tidymodels_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mgghighlight\u001b[39m::\u001b[32maes()\u001b[39m       masks \u001b[34mggtern\u001b[39m::aes(), \u001b[34mggplot2\u001b[39m::aes()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mannotate()\u001b[39m       masks \u001b[34mggplot2\u001b[39m::annotate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mbetween()\u001b[39m         masks \u001b[34mdata.table\u001b[39m::between()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mrecipes\u001b[39m::\u001b[32mcheck()\u001b[39m         masks \u001b[34mpermute\u001b[39m::check()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdials\u001b[39m::\u001b[32mdegree()\u001b[39m          masks \u001b[34mcirclize\u001b[39m::degree()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mscales\u001b[39m::\u001b[32mdiscard()\u001b[39m        masks \u001b[34mpurrr\u001b[39m::discard()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mMatrix\u001b[39m::\u001b[32mexpand()\u001b[39m         masks \u001b[34mtidyr\u001b[39m::expand()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mextract()\u001b[39m         masks \u001b[34mmagrittr\u001b[39m::extract()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m          masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfirst()\u001b[39m           masks \u001b[34mdata.table\u001b[39m::first()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mrecipes\u001b[39m::\u001b[32mfixed()\u001b[39m         masks \u001b[34mstringr\u001b[39m::fixed()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mflatten()\u001b[39m         masks \u001b[34mjsonlite\u001b[39m::flatten()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mggplot_build()\u001b[39m   masks \u001b[34mggplot2\u001b[39m::ggplot_build()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mggplot_gtable()\u001b[39m  masks \u001b[34mggplot2\u001b[39m::ggplot_gtable()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mggsave()\u001b[39m         masks \u001b[34mggplot2\u001b[39m::ggsave()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m             masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlast()\u001b[39m            masks \u001b[34mdata.table\u001b[39m::last()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mlayer_data()\u001b[39m     masks \u001b[34mggplot2\u001b[39m::layer_data()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mcaret\u001b[39m::\u001b[32mlift()\u001b[39m            masks \u001b[34mpurrr\u001b[39m::lift()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdials\u001b[39m::\u001b[32mmixture()\u001b[39m         masks \u001b[34mbrms\u001b[39m::mixture()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mvegan\u001b[39m::\u001b[32mnullmodel()\u001b[39m       masks \u001b[34mparsnip\u001b[39m::nullmodel()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mMatrix\u001b[39m::\u001b[32mpack()\u001b[39m           masks \u001b[34mtidyr\u001b[39m::pack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mrsample\u001b[39m::\u001b[32mpopulate()\u001b[39m      masks \u001b[34mRcpp\u001b[39m::populate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34myardstick\u001b[39m::\u001b[32mprecision()\u001b[39m   masks \u001b[34mcaret\u001b[39m::precision()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mrecipes\u001b[39m::\u001b[32mprepare()\u001b[39m       masks \u001b[34mComplexHeatmap\u001b[39m::prepare()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdials\u001b[39m::\u001b[32mprune()\u001b[39m           masks \u001b[34mdendextend\u001b[39m::prune()\n",
      "\u001b[31m✖\u001b[39m \u001b[34myardstick\u001b[39m::\u001b[32mrecall()\u001b[39m      masks \u001b[34mcaret\u001b[39m::recall()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mMASS\u001b[39m::\u001b[32mselect()\u001b[39m           masks \u001b[34mdplyr\u001b[39m::select()\n",
      "\u001b[31m✖\u001b[39m \u001b[34myardstick\u001b[39m::\u001b[32msensitivity()\u001b[39m masks \u001b[34mcaret\u001b[39m::sensitivity()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mset_names()\u001b[39m       masks \u001b[34mmagrittr\u001b[39m::set_names()\n",
      "\u001b[31m✖\u001b[39m \u001b[34myardstick\u001b[39m::\u001b[32mspec()\u001b[39m        masks \u001b[34mreadr\u001b[39m::spec()\n",
      "\u001b[31m✖\u001b[39m \u001b[34myardstick\u001b[39m::\u001b[32mspecificity()\u001b[39m masks \u001b[34mcaret\u001b[39m::specificity()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mrecipes\u001b[39m::\u001b[32mstep()\u001b[39m          masks \u001b[34mstats\u001b[39m::step()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_bw()\u001b[39m       masks \u001b[34mggplot2\u001b[39m::theme_bw()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_classic()\u001b[39m  masks \u001b[34mggplot2\u001b[39m::theme_classic()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_dark()\u001b[39m     masks \u001b[34mggplot2\u001b[39m::theme_dark()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_gray()\u001b[39m     masks \u001b[34mggplot2\u001b[39m::theme_gray()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_light()\u001b[39m    masks \u001b[34mggplot2\u001b[39m::theme_light()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_linedraw()\u001b[39m masks \u001b[34mggplot2\u001b[39m::theme_linedraw()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_minimal()\u001b[39m  masks \u001b[34mggplot2\u001b[39m::theme_minimal()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mggtern\u001b[39m::\u001b[32mtheme_void()\u001b[39m     masks \u001b[34mggplot2\u001b[39m::theme_void()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mtranspose()\u001b[39m       masks \u001b[34mdata.table\u001b[39m::transpose()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mMatrix\u001b[39m::\u001b[32munpack()\u001b[39m         masks \u001b[34mtidyr\u001b[39m::unpack()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mrecipes\u001b[39m::\u001b[32mupdate()\u001b[39m        masks \u001b[34mMatrix\u001b[39m::update(), \u001b[34mstats\u001b[39m::update()\n",
      "\u001b[34m•\u001b[39m Dig deeper into tidy modeling with R at \u001b[32mhttps://www.tmwr.org\u001b[39m\n",
      "\n",
      "\n",
      "Attaching package: ‘performance’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:yardstick’:\n",
      "\n",
      "    mae, rmse\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:arm’:\n",
      "\n",
      "    display\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘AICcPermanova’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:fmsb’:\n",
      "\n",
      "    VIF\n",
      "\n",
      "\n",
      "Welcome to compositions, a package for compositional data analysis.\n",
      "Find an intro with \"? compositions\"\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘compositions’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:yardstick’:\n",
      "\n",
      "    msd\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:rstanarm’:\n",
      "\n",
      "    R2\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:arm’:\n",
      "\n",
      "    balance\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Matrix’:\n",
      "\n",
      "    norm\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:caret’:\n",
      "\n",
      "    R2\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:arrow’:\n",
      "\n",
      "    binary, scalar\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    anova, cor, cov, dist, var\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:graphics’:\n",
      "\n",
      "    segments\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    %*%, norm, scale, scale.default\n",
      "\n",
      "\n",
      "Loading required package: cluster\n",
      "\n"
     ]
    }
   ],
   "source": [
    "source('R_functions/cdr3-QTL_functions.R')\n",
    "source('R_functions//hla_functions.R')\n",
    "source('libraries.R')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Making matrices for MANOVA and Conditional in DOWNSAMPLED CDR3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true,
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "hla_genotypes_matrix_wo_low_ab <- fread('../data/hla/hla_genotypes_matrix_wo_low_ab.csv')\n",
    "pca_hla <- fread('../data/hla/pca_hla_9PCs.tsv')\n",
    "hla_matrices <- paste0('../hla_matrices/with_pcs/',list.files('../hla_matrices/with_pcs/'))\n",
    "pcs <- 3\n",
    "dir_downsampling <- '../CDR3_downsampling_experiment/unique/'\n",
    "dir_content <- paste0(dir_downsampling, list.files(dir_downsampling))\n",
    "\n",
    "for (path_in in dir_content[2:length(dir_content)]){\n",
    "    sample_size <- unlist(strsplit(path_in, '/'))[4]\n",
    "    source(\"Make_matrices_for_Main_Manova.R\", chdir = TRUE)\n",
    "    \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "path_in <- \"/work_beegfs/sukmb667/projects/cdr3-qtl/healthy_and_ibd/CDR3_downsampling_experiment/unique/\"\n",
    "directories <- list.files(path_in, full.names = TRUE)\n",
    "pairs_dir <- directories[file.info(directories)$isdir]\n",
    "\n",
    "for (tsv_dir in pairs_dir) {\n",
    "    sample_size <- strsplit(tsv_dir, \"/\")[[1]][length(strsplit(tsv_dir, \"/\")[[1]])]\n",
    "    # List all .tsv files in the directory\n",
    "    tsv_files <- list.files(paste0(tsv_dir, '/manova_results/cdr3_hla_pairs/irt_freq_unique/'), pattern = \"\\\\.tsv$\", full.names = TRUE)\n",
    " \n",
    "    # Read all .tsv files into a list of data.tables\n",
    "    data_table_list <- lapply(tsv_files, fread)\n",
    "\n",
    "    save(data_table_list, file = paste0(tsv_dir,'/',sample_size,\"_irt_freq_unique.RData\"))\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "### Main Manova on CDR3 Downsampled"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "source('/work_beegfs/sukmb667/projects/cdr3-qtl/healthy_and_ibd/scripts/Main_manova_from_matrices.R', chdir = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Export .db to the .csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "db_path <- \"/work_beegfs/sukmb667/projects/cdr3-qtl/healthy_and_ibd/CDR3_downsampling_experiment/unique/cdr3-downsampling-unique.db\"\n",
    "conn <- dbConnect(RSQLite::SQLite(), db_path)\n",
    "tables <- dbListTables(conn)\n",
    "df <- dbReadTable(conn,  tables)\n",
    "\n",
    "output_path <- \"/work_beegfs/sukmb667/projects/cdr3-qtl/healthy_and_ibd/CDR3_downsampling_experiment/unique/Downsampling_manova_all.csv\"\n",
    "fwrite(df, output_path, row.names = FALSE)\n",
    "\n",
    "dbDisconnect(conn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "first <- TRUE\n",
    "sample_size <- '100000'\n",
    "path_pairs <- paste0(path_in, '/manova_results/cdr3_hla_pairs/irt_freq_unique/')\n",
    "pairs_to_test <- paste0(path_pairs, list.files(path_pairs))\n",
    "#pairs_to_test <- grep(paste(tested_pairs, collapse = \"|\"), list.files(path_pairs), value = TRUE, invert = TRUE, fixed = TRUE)\n",
    "for (pair in pairs_to_test){\n",
    "    pair_matrix <- fread(pair)\n",
    "    dir_results <- paste0(path_in, '/manova_results/irt_freq_unique/')\n",
    "    manova_df <- mlm_fun(pair_matrix, dir_results, n_pcs = pcs) %>% \n",
    "        separate(pair, into = c('HLA', 'Site_hla', 'Length_cdr3', 'Position_cdr3'), sep = '_', remove = FALSE)\n",
    "    if (first){\n",
    "        manova_df_all <- manova_df[0,]\n",
    "        first <- FALSE\n",
    "    } else {\n",
    "    manova_df_all <- rbind(manova_df_all, manova_df) \n",
    "        }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conditional on Downsampled CDR3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "hla_matrices <- paste0('../hla_matrices/with_pcs/',list.files('../hla_matrices/with_pcs/'))\n",
    "pcs <- 3\n",
    "\n",
    "for (path_in in dir_content{\n",
    "    dir_analysis <- paste0(path_in, '/conditional_analysis/')\n",
    "    if (!file.exists(dir_analysis)) {\n",
    "        dir.create(dir_analysis, recursive = TRUE)\n",
    "        }\n",
    "    pairs_to_test <- paste0(path_in,'/cdr3_hla_pairs/')\n",
    "    sample_size <- unlist(strsplit(path_in, '/'))[3]\n",
    "    manova_cond_all <- fread(paste0(path_in,'/',sample_size,'_manova.tsv'))\n",
    "    bonf <- 0.01 / nrow(na.omit(manova_cond_all))\n",
    "            \n",
    "    source('Conditional_analysis_updated.R')\n",
    "                 }"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
